Linker on OSX x64 generates errors

## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=1109875&view=logs&j=2bc6bdce-e470-5dcb-3197-fe2ccb7d0240
Build error leg or test failing: Build / osx-x64 Release NativeAOT_Libs / Build product
Pull request: https://github.com/dotnet/runtime/pull/118211
<!-- Error message template  -->
## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

<!-- Use ErrorMessage for String.Contains matches. Use ErrorPattern for regex matches (single line/no backtracking). Set BuildRetry to `true` to retry builds with this error. Set ExcludeConsoleLog to `true` to skip helix logs analysis. -->

```json
{
  "ErrorMessage": "unexpected end of data at offset",
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```



E.g. https://dev.azure.com/dnceng-public/public/_build/results?buildId=1109875&view=logs&jobId=2bc6bdce-e470-5dcb-3197-fe2ccb7d0240&j=2bc6bdce-e470-5dcb-3197-fe2ccb7d0240&t=e26b3874-8455-5b0e-bfa1-ab355ae88e1e

I assume we updated macOS or XCode on the build machines.

```
EXEC : error : unexpected end of data at offset 0x110157 while reading [0x10cf79, 0x11907a) [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyModel/tests/Microsoft.Extensions.DependencyModel.Tests.csproj::TargetFramework=net10.0]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) unexpected end of data at offset 0x110157 while reading [0x10cf79, 0x11907a)
EXEC : error : last sequence in debug line table at offset 0x00000000 is not terminated [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyModel/tests/Microsoft.Extensions.DependencyModel.Tests.csproj::TargetFramework=net10.0]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) last sequence in debug line table at offset 0x00000000 is not terminated
EXEC : error : unexpected end of data at offset 0x110157 while reading [0x10cf79, 0x11907a) [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyModel/tests/Microsoft.Extensions.DependencyModel.Tests.csproj::TargetFramework=net10.0]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) unexpected end of data at offset 0x110157 while reading [0x10cf79, 0x11907a)
EXEC : error : last sequence in debug line table at offset 0x00000000 is not terminated [/Users/runner/work/1/s/src/libraries/Microsoft.Extensions.DependencyModel/tests/Microsoft.Extensions.DependencyModel.Tests.csproj::TargetFramework=net10.0]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) last sequence in debug line table at offset 0x00000000 is not terminated
```
<!-- Known issue validation start -->
 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=1109875
**Error message validated:** `[unexpected end of data at offset`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 7/31/2025 10:44:38 AM UTC
<!-- Known issue validation end -->
<!--Known issue error report start -->

### Report
|Build|Definition|Step Name|Console log|Pull Request|
|---|---|---|---|---|
|[1121395](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1121395)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1121395/logs/154)|dotnet/runtime#118676|
|[1121318](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1121318)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1121318/logs/191)|dotnet/runtime#118673|
|[1121265](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1121265)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1121265/logs/171)||
|[1120868](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1120868)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1120868/logs/214)||
|[1120257](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1120257)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1120257/logs/201)|dotnet/runtime#118632|
|[1119756](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1119756)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1119756/logs/140)||
|[1119238](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1119238)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1119238/logs/181)||
|[1118934](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118934)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118934/logs/138)||
|[1118829](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118829)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118829/logs/181)||
|[1118740](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118740)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118740/logs/140)||
|[1118624](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118624)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118624/logs/159)||
|[1118380](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118380)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118380/logs/175)|dotnet/runtime#118528|
|[1118286](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1118286)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1118286/logs/191)||
|[1117610](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1117610)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1117610/logs/204)||
|[1117472](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1117472)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1117472/logs/182)|dotnet/runtime#117502|
|[1117151](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1117151)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1117151/logs/144)|dotnet/runtime#117502|
|[1117079](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1117079)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1117079/logs/173)||
|[1116603](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1116603)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1116603/logs/211)||
|[1116100](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1116100)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1116100/logs/202)||
|[1115255](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1115255)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1115255/logs/244)||
|[1114914](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1114914)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1114914/logs/172)|dotnet/runtime#111178|
|[1114687](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1114687)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1114687/logs/247)||
|[1113990](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1113990)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1113990/logs/256)||
|[1113884](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1113884)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1113884/logs/156)|dotnet/runtime#111178|
|[1113425](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1113425)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1113425/logs/148)|dotnet/runtime#118324|
|[1113409](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1113409)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1113409/logs/200)||
|[1112650](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1112650)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1112650/logs/139)||
|[1112366](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1112366)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1112366/logs/140)||
|[1112206](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1112206)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1112206/logs/152)|dotnet/runtime#118060|
|[1112203](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1112203)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1112203/logs/160)||
|[1112097](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1112097)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1112097/logs/139)||
|[1111948](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111948)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111948/logs/148)||
|[1111702](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111702)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111702/logs/155)|dotnet/runtime#118256|
|[1111635](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111635)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111635/logs/136)||
|[1111302](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111302)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111302/logs/121)|dotnet/runtime#118273|
|[1111300](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111300)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111300/logs/165)|dotnet/runtime#118060|
|[1111090](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111090)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111090/logs/159)||
|[1111072](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111072)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111072/logs/145)|dotnet/runtime#118156|
|[1111000](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1111000)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1111000/logs/126)|dotnet/runtime#118060|
|[1110967](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110967)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110967/logs/119)|dotnet/runtime#118256|
|[1110709](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110709)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110709/logs/212)||
|[1110663](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110663)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110663/logs/154)|dotnet/runtime#118235|
|[1110287](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110287)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110287/logs/155)|dotnet/runtime#118235|
|[1110271](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110271)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110271/logs/114)|dotnet/runtime#118231|
|[1110198](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1110198)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1110198/logs/192)||
|[1109875](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1109875)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/1109875/logs/154)|dotnet/runtime#118211|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|3|18|46|
<!--Known issue error report end -->

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
@@ -517,6 +517,13 @@ namespace ILCompiler.ObjectWriter
                         continue;
                     }
 
+                    ISymbolNode symbolNode = node as ISymbolNode;
+                    ISymbolNode deduplicatedSymbolNode = _nodeFactory.ObjectInterner.GetDeduplicatedSymbol(_nodeFactory, symbolNode);
+                    if (deduplicatedSymbolNode != symbolNode)
+                    {
+                        continue;
+                    }
+
                     // Ensure any allocated MethodTables have debug info
                     if (node is ConstructedEETypeNode methodTable)
                     {

